---
title: "Results integration"
output: pdf_document
date: "2023-12-21"
---

```{r,  message = FALSE, results='hide', warning=FALSE}
# Set the CRAN mirror
options(repos = c(CRAN = "https://cran.rstudio.com"))
install.packages("gplots")
install.packages("dplyr")
library(dplyr)
library(gplots)
library(tidyr)
```


# Integrating the results of Differential Gene Expression Analysis
Loading the results of the analysis of the different datasets
```{r}
DGE1 <- read.table("Dataset1_limma_results.txt", header = TRUE, sep = "\t")
DGE2 <- read.table("Dataset2_limma_results.txt", header = TRUE, sep = "\t")
DGE3 <- read.table("Dataset3_limma_results.txt", header = TRUE, sep = "\t")
DGE4 <- read.table("Dataset4_limma_results.txt", header = TRUE, sep = "\t")
head(DGE1)
head(DGE2)
head(DGE3)
head(DGE4)
```

Only the columns of interest are kept, and columnnames are unified. Rows that have no gene symbol are omitted from the dataframe.
```{r}
DGE1 <- DGE1[, c("logFC", "adj.P.Val", "Symbol")]
DGE1 <- DGE1[complete.cases(DGE1$Symbol), ]
DGE1 <- DGE1[DGE1$Symbol != "", , drop = FALSE]

DGE2$Symbol <- DGE2$gene_symbol
DGE2 <- DGE2[, c("logFC", "adj.P.Val", "Symbol")]
DGE2 <- DGE2[complete.cases(DGE2$Symbol), ]
DGE2 <- DGE2[DGE2$Symbol != "", , drop = FALSE]

DGE3$Symbol <- DGE3$gene
DGE3 <- DGE3[, c("logFC", "adj.P.Val", "Symbol")]
DGE3 <- DGE3[complete.cases(DGE3$Symbol), ]
DGE3 <- DGE3[DGE3$Symbol != "", , drop = FALSE]

DGE4$Symbol <- DGE4$Genes
DGE4 <- DGE4[, c("logFC", "adj.P.Val", "Symbol")]
DGE4 <- DGE4[complete.cases(DGE4$Symbol), ]
DGE4 <- DGE4[DGE4$Symbol != "", , drop = FALSE]

head(DGE1)
head(DGE2)
head(DGE3)
head(DGE4)
```

The third dataset showed the highest number of significantly differentially expressed genes.
```{r}
dim(DGE3[DGE3$adj.P.Val < 0.05, ])
```

The first 5112 genes are selected from each dataset, from which respectively upregulated and downregulated genes are used.
```{r}
DGE1_comp <- DGE1[1:5112, ]
DGE2_comp <- DGE2[1:5112, ]
DGE3_comp <- DGE3[1:5112, ]
DGE4_comp <- DGE4[1:5112, ]

DGE1_up <- DGE1_comp$Symbol[DGE1_comp$logFC > 0]
DGE2_up <- DGE2_comp$Symbol[DGE2_comp$logFC > 0]
DGE3_up <- DGE3_comp$Symbol[DGE3_comp$logFC > 0]
DGE4_up <- DGE4_comp$Symbol[DGE4_comp$logFC > 0]

DGE1_down <- DGE1_comp$Symbol[DGE1_comp$logFC < 0]
DGE2_down <- DGE2_comp$Symbol[DGE2_comp$logFC < 0]
DGE3_down <- DGE3_comp$Symbol[DGE3_comp$logFC < 0]
DGE4_down <- DGE4_comp$Symbol[DGE4_comp$logFC < 0]
```

# Upregulated Genes Overlap from subset
```{r}
# Extract unique symbols from each set of upregulated genes
symbols_up <- list(DGE1 = unique(DGE1_up),
                   DGE2 = unique(DGE2_up),
                   DGE3 = unique(DGE3_up),
                   DGE4 = unique(DGE4_up))

# Create Venn diagram for upregulated genes
venn <- venn(symbols_up)

```

There are 8 overlapping upregulated genes found.
```{r}
# Extract unique symbols from each upregulated dataframe
symbols_DGE1_up <- unique(DGE1_up)
symbols_DGE2_up <- unique(DGE2_up)
symbols_DGE3_up <- unique(DGE3_up)
symbols_DGE4_up <- unique(DGE4_up)

# Find common symbols among upregulated genes
common_symbols_up <- Reduce(intersect, list(symbols_DGE1_up, symbols_DGE2_up, symbols_DGE3_up, symbols_DGE4_up))

# Print or use common_symbols_up as needed
print(common_symbols_up)
```

The adj.p.value and logFC of each of the genes is checked for each dataset
```{r}
# Create a vector with the overlapping genes
overlapping_genes <- c("TGOLN2", "CRYZ", "SQSTM1", "ALG5", "BLVRA", "ST3GAL1", "CAMLG", "SNCA")

# Function to extract relevant columns for each dataset
extract_columns <- function(dataset, genes) {
  result <- dataset[dataset$Symbol %in% genes, c("Symbol", "adj.P.Val", "logFC")]
  
  # Convert "logFC" to numeric, as it might contain non-numeric values
  result$logFC <- as.numeric(as.character(result$logFC))
  
  colnames(result) <- c("Symbol", "adj.p.value", "logFC")
  
 # Keep the row with the maximum absolute logFC for each unique gene symbol
  result <- result %>%
    group_by(Symbol) %>%
    filter(logFC == max(logFC))
  
  return(result)
}

# Extract columns for each dataset
DGE1_subset <- extract_columns(DGE1, overlapping_genes)
DGE2_subset <- extract_columns(DGE2, overlapping_genes)
DGE3_subset <- extract_columns(DGE3, overlapping_genes)
DGE4_subset <- extract_columns(DGE4, overlapping_genes)


# Combine the four subsets into a single dataframe
combined_df <- bind_rows(
  mutate(DGE1_subset, Dataset = "Dataset 1"),
  mutate(DGE2_subset, Dataset = "Dataset 2"),
  mutate(DGE3_subset, Dataset = "Dataset 3"),
  mutate(DGE4_subset, Dataset = "Dataset 4"))

sorted_df <- combined_df %>% arrange(Symbol)

sorted_df
```



# Downregulated Genes Overlap from subset
```{r}
# Extract unique symbols from each set of upregulated genes
symbols_down <- list(DGE1 = unique(DGE1_down),
                   DGE2 = unique(DGE2_down),
                   DGE3 = unique(DGE3_down),
                   DGE4 = unique(DGE4_down))

# Create Venn diagram for upregulated genes
venn <- venn(symbols_down)
```

There are 2 downregulated genes found.
```{r}
# Extract unique symbols from each downregulated dataframe
symbols_DGE1_down <- unique(DGE1_down)
symbols_DGE2_down <- unique(DGE2_down)
symbols_DGE3_down <- unique(DGE3_down)
symbols_DGE4_down <- unique(DGE4_down)

# Find common symbols among downregulated genes
common_symbols_down <- Reduce(intersect, list(symbols_DGE1_down, symbols_DGE2_down, symbols_DGE3_down, symbols_DGE4_down))

# Print or use common_symbols_down as needed
print(common_symbols_down)
```

The adj.p.value and logFC of each of the genes is checked for each dataset
```{r}
# Create a vector with the overlapping genes
overlapping_genes <- c("DEPDC1", "NTRK3")

# Function to extract relevant columns for each dataset
extract_columns <- function(dataset, genes) {
  result <- dataset[dataset$Symbol %in% genes, c("Symbol", "adj.P.Val", "logFC")]
  
  # Convert "logFC" to numeric, as it might contain non-numeric values
  result$logFC <- as.numeric(as.character(result$logFC))
  
  colnames(result) <- c("Symbol", "adj.p.value", "logFC")
  
 # Keep the row with the maximum absolute logFC for each unique gene symbol
  result <- result %>%
    group_by(Symbol) %>%
    filter(logFC == min(logFC))
  
  return(result)
}

# Extract columns for each dataset
DGE1_subset <- extract_columns(DGE1, overlapping_genes)
DGE2_subset <- extract_columns(DGE2, overlapping_genes)
DGE3_subset <- extract_columns(DGE3, overlapping_genes)
DGE4_subset <- extract_columns(DGE4, overlapping_genes)


# Combine the four subsets into a single dataframe
combined_df <- bind_rows(
  mutate(DGE1_subset, Dataset = "Dataset 1"),
  mutate(DGE2_subset, Dataset = "Dataset 2"),
  mutate(DGE3_subset, Dataset = "Dataset 3"),
  mutate(DGE4_subset, Dataset = "Dataset 4"))

sorted_df <- combined_df %>% arrange(Symbol)

sorted_df
```