---
title: 'Analysis of Dataset 2 -  Expression profiling by High Throughput Sequencing (GSE216609)'
output: pdf_document
date: "2023-12-22"
---
# Data Exploration and Preprocessing
```{r,  message = FALSE, results='hide', warning=FALSE}
# Set the CRAN mirror
options(repos = c(CRAN = "https://cran.rstudio.com"))
## Install packages
BiocManager::install("edgeR", update = F)
BiocManager::install("DESeq2", update = F)
BiocManager::install("limma", update = F)
BiocManager::install("tximport", update = F)
BiocManager::install("tximeta", update = F)
BiocManager::install("biomaRt", update = F)

## Load packages
library(edgeR)
library(DESeq2)
library(limma)
library(tximport)
library(tximeta)
library(biomaRt)

## Set working directory
setwd("C:/Users/maxim/OneDrive - UGent/1 Master Bio-informatics/Semester 1/Applied High Throughput Analysis/Project/Dataset 2 - GSE216609/HiSeqData")
```

Transcript annotation is loaded to be merged with the kallisto results.

```{r}
# Get annotation data
grch38 <-  useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")

data <- getBM(attributes = c('ensembl_gene_id', 'ensembl_transcript_id', 'external_gene_name', 'entrezgene_id', 'transcript_length'), mart = grch38)

tx2gene <- dplyr::select(data, ensembl_transcript_id, ensembl_gene_id, external_gene_name, entrezgene_id, transcript_length)
tx2gene <- dplyr::rename(tx2gene, TXNAME = ensembl_transcript_id)
tx2gene <- dplyr::rename(tx2gene, GENEID = ensembl_gene_id)
tx2gene <- dplyr::rename(tx2gene, GENENAME = external_gene_name)
tx2gene <- dplyr::rename(tx2gene, ENTREZID = entrezgene_id)

head(tx2gene)
```

Loading Kallisto results.

```{r}
# Get file locations
files <- list.files("kallisto_quant", pattern=".tsv")
files <- files[grep("abundance.tsv",files)]
samples <- unlist(strsplit(files,"_"))[c(1:length(files))*2-1]
files <- paste0("Kallisto_quant/", files)
names(files) <- samples

# Load RNAseq data; normalization for transcript length and library size
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene, countsFromAbundance = "lengthScaledTPM")

head(txi$counts)
dim(txi$counts)
```

Checking for duplicate rows
```{r}
sum(duplicated(rownames(txi$counts)))
```

Merging metatdata with the results. There are no confounders present in the study.
```{r}
sdrf_rnaseq <- read.delim("SraRunTable.txt", sep=',')
print(sdrf_rnaseq[,c("Run","cell_type","disease_state","sex")])
```

## Preprocessing

```{r}
# Creating a DGEList object for use in edgeR
y <- DGEList(txi$counts)

# Filtering
condition <- factor(sdrf_rnaseq$disease_state)
y$samples$group <- condition
keep <- filterByExpr(y, group = y$samples$group)
y <- y[keep, , keep.lib.sizes=FALSE]
summary(keep)
```
Calculating TMM normalisation factors
```{r}
y <- calcNormFactors(y)
y$samples
```

PCA on the normalized count data to visualize the overall grouping of samples
```{r}
plotMDS(y)
```


Making design
```{r}

design <- model.matrix(~0 + condition)
colnames(design) <- c('control', 'PCOS')
design
```

# Differential expression analysis

```{r}
v <- voomWithQualityWeights(y, design, plot=TRUE)
```
```{r}
v
```

```{r}
fit_limma <- lmFit(v, design)
cont.matrix <- makeContrasts(PCOS - control, levels = design)
fit2 <- contrasts.fit(fit_limma, cont.matrix)
fit_limma2 <- eBayes(fit2)
res_limma <- topTable(fit_limma2, coef=1, adjust.method="BH", number=Inf)
res_limma
```

```{r}
topGenes <- res_limma[res_limma$adj.P.Val < 0.05 & abs(res_limma$logFC) < 1.5, ]
topGenes
```


MA plot
```{r}
jpeg("res_limmaWQW_PCOS_MA.png")
limma::plotMA(fit_limma2)
dev.off()
limma::plotMA(fit_limma2)
```

Volcano plot
```{r}
jpeg("res_limmaWQW_PCOS_volcano.png")
limma::volcanoplot(fit_limma2, coef=1)
dev.off()
limma::volcanoplot(fit_limma2, coef=1)
```

# Gene Set Analysis

## Gene annotation
As the package WebGestaltR is used, which requires EntrezIDs or Ensembl IDs to perform an Over Representation Analysis, annotation of the genes is performed first.
```{r}
# Extract gene names from tx2gene
gene_names <- tx2gene$GENENAME

# Match gene names based on gene IDs
res_limma$GENENAME <- gene_names[match(rownames(res_limma), tx2gene$GENEID)]

# Match gene IDs based on gene names
gene_ids <- tx2gene$GENEID[match(res_limma$GENENAME, tx2gene$GENENAME)]
transcript_ids <- tx2gene$TXNAME[match(res_limma$GENENAME, tx2gene$GENENAME)]
transcript_lengths <- tx2gene$transcript_length[match(res_limma$GENENAME, tx2gene$GENENAME)]

# Add ENSEMBL_GENE_ID and ENSEMBL_TRANSCRIPT_ID to res_limma
res_limma$ensembl_gene_id <- gene_ids
res_limma$ensembl_transcript_id <- transcript_ids

# Add transcript to res_limma
res_limma$transcript_length <- transcript_lengths

# Display the annotated result
res_limma
```


```{r}
# Sort results based on Ensembl gene ID
res_sorted <- res_limma[order(res_limma$ensembl_gene_id), ]
head(res_sorted)
```


```{r}
# Check dimensions and overlap
cat("Overlap of Ensembl gene IDs between results and annotation data:",
    sum(res_sorted$ensembl_gene_id %in% tx2gene$GENEID), "\n")
cat("Dimensions of sorted results:", dim(res_sorted), "\n")
cat("Dimensions of annotation data:", dim(tx2gene), "\n")
```


```{r}
# Add annotation to sorted results
res_sorted$gene_symbol <- tx2gene$GENENAME[match(res_sorted$ensembl_gene_id, tx2gene$GENEID)]
res_sorted$EntrezID <- tx2gene$ENTREZID[match(res_sorted$ensembl_gene_id, tx2gene$GENEID)]
res_sorted$transcript_length <- tx2gene$transcript_length[match(res_sorted$ensembl_gene_id, tx2gene$GENEID)]
```


```{r}
# Sort on p-values
res_annot <- res_sorted[order(res_sorted$P.Value), ]
head(res_annot)
```


## Over Representation Analysis
A filtering step is performed to only have entries with EntrezIDs

```{r}
res_annot <- res_annot[complete.cases(res_annot$EntrezID), ]
```


The top 300 genes are selected based on p-value and logFC, these genes are most likely to be differently expressed.

```{r}
logFC_interest <- res_annot[abs(res_annot$logFC) > 1.5, ]
logFC_interest_sorted <- logFC_interest[sort(logFC_interest$P.Value, index.return = T)$ix, ]
top_genes <- head(logFC_interest_sorted, 300)
head(top_genes)
```

From these genes, we select the upregulated and downregulated genes. We store the entrez ids of these genes in txt files for further use in the Over Representation Analysis. The reference gene set is equal to all genes in the array.

```{r}
# Upregulated genes
upreg_genes <- top_genes$EntrezID[top_genes$logFC > 0]
upreg_genes <- upreg_genes[!is.na(upreg_genes)]
cat("Number of downregulated genes:", length(upreg_genes), "\n")

# Downregulated genes
downreg_genes <- top_genes$EntrezID[top_genes$logFC < 0]
downreg_genes <- downreg_genes[!is.na(downreg_genes)]
cat("Number of downregulated genes:", length(downreg_genes), "\n")

# Reference genes
ref_genes <- res_annot$EntrezID[!is.na(res_annot$EntrezID)]
cat("Number of reference genes:", length(ref_genes), "\n")
```

### KEGG

```{r}
# Perform KEGG pathway analysis on the upregulated genes
kegg_upreg <- limma::kegga(de = upreg_genes, universe = ref_genes, trend = res_annot$transcript_length, species = "Hs")

# Sort results and calculate FDR
kegg_upreg <- kegg_upreg[sort(kegg_upreg$P.DE, index.return=T)$ix, ]
kegg_upreg$P.DE.adj <- p.adjust(kegg_upreg$P.DE, n=nrow(kegg_upreg), method="BH")
head(kegg_upreg)
```


```{r}
# Perform KEGG pathway analysis on the downregulated genes
kegg_downreg <- limma::kegga(de = downreg_genes, universe = ref_genes, trend = res_annot$transcript_length, species = "Hs")

# Print or further process the results
# Sort results and calculate FDR
kegg_downreg <- kegg_downreg[sort(kegg_downreg$P.DE, index.return=T)$ix, ]
kegg_downreg$P.DE.adj <- p.adjust(kegg_downreg$P.DE, n=nrow(kegg_downreg), method="BH")
head(kegg_downreg)
```

### Gene Ontology: Molecular Function (MF)

```{r}
# Perform Gene Ontology: Molecular Function Analysis on the upregulated genes
goana_upreg <- limma::goana(de = upreg_genes, universe = ref_genes, trend = res_annot$transcript_length, species = "Hs")

# Print or further process the results
goana_upreg <- goana_upreg[goana_upreg$Ont == "MF", ]
goana_upreg <- goana_upreg[sort(goana_upreg$P.DE, index.return=T)$ix, ]
goana_upreg$P.DE.adj <- p.adjust(goana_upreg$P.DE, n=nrow(goana_upreg), method="BH")
head(goana_upreg)
```


```{r}
# Perform Gene Ontology: Molecular Function Analysis using WebGestaltR on the downregulated genes
goana_downreg <- limma::goana(de = downreg_genes, universe = ref_genes, trend = res_annot$transcript_length, species = "Hs")

# Print or further process the results
goana_downreg <- goana_downreg[goana_downreg$Ont == "MF", ]
goana_downreg <- goana_downreg[sort(goana_downreg$P.DE, index.return=T)$ix, ]
goana_downreg$P.DE.adj <- p.adjust(goana_downreg$P.DE, n=nrow(goana_downreg), method="BH")
head(goana_downreg)
```


# Writing out Data for comparison

Results of the analysis of Dataset 1 are saved for comparison between dataset analysis.
The results of limma analysis are saved in txt file for further use.

```{r}
write.table(res_annot, sep= "\t", file="Dataset2_limma_results.txt")
```

The results of gene set analysis are saved in a txt file for further use.

```{r}
write.table(kegg_upreg, sep= "\t", file="Dataset2_PathwayAnalysis_upreg_results.txt")
write.table(kegg_downreg, sep= "\t", file="Dataset2_PathwayAnalysis_downreg_results.txt")
write.table(goana_upreg, sep= "\t", file="Dataset2_MolecularFunction_upreg_results.txt")
write.table(goana_downreg, sep= "\t", file="Dataset2_MolecularFunction_downreg_results.txt")
```
